### ISSUE TITLE
MP4Client unable to play SHVC-encoded video

### ISSUE THREAD

**Issue by bayantaani** [gpac/gpac#865]
Hello,
I have two MP4 videos encoded using SHVC reference software SHM12.3; one with one layer only (out.mp4) and the second with 2 layers (out2l.mp4).
MP4Client is able to play the first version, but unable to play the second (with 2 layers).
The command I used is `MP4Client out.mp4` for the first version, and although it plays out the video, I get the following error in terminal:
`[Compositor2D] Cannot attach video surface memory to raster: Feature Not Supported`

For the second version: `MP4Client out2l.mp4` the video doesn't play at all, and I get the following in the terminal:
`[Codec] ODM1 ES2: Cannot find decoder for stream type MPEG-H L-HEVC Video
 Stream Setup Failure: Media Codec not found
[Compositor2D] Cannot attach video surface memory to raster: Feature Not Supported`

Both videos are encoded using the same SHM12.3 software, the only difference is with the number of layers.
The two videos can be downloaded from [this link](https://drive.google.com/file/d/0BzO-c87ZZxWAUE53dGNFM3pCTEE/view?usp=sharing)

How can I add SHVC decoder to MP4Client? And why does this error show only when I have a video of 2 layers?

P.S. I can play both videos on VLC player

**Comment by aureliendavid** [comment]
I've looked into this a bit, and here's my understanding: 

 * there are two hevc decoding modules in gpac: ffmpeg and openhevc
 * ffmpeg has support for HEVC but not SHVC
 * openhevc has (some) support for SHVC
 * your first video is actually plain HEVC so it will play with both
 * the second uses SHVC extensions and will only play with the openhevc module

so far so good, but: 

 * openhevc is only available on 64 bits architectures (shouldn't be a problem nowadays)
 * there is a problem in our automatic builds which is currently preventing openhevc from being enabled in our linux-64 binaries (this is an easy fix and I'm doing it right now)
 * the branch of openhevc we're using on windows crashes with your video, I'm looking into using a better one
 * I haven't looked at OSX yet

**TL;DR:** if you're using linux 64 bits, try this binary ; https://download.tsi.telecom-paristech.fr/gpac/latest_builds/linux64/gpac/gpac_0.7.2-DEV-rev176-g2810ba5-master_amd64.deb (if you're not it might take longer to fix)

**Comment by bayantaani** [comment]
Thank you very much for the binary, it gave me a segmentation fault in the beginning but then I ran it again and now I can play both videos without any problems.
BTW I am mainly using OSX, that's why it took me some time to install a Linux machine and run it. It would be great if you also get it fixed for OSX.

Thanks!

**Comment by aureliendavid** [comment]
I think it should now be ok on OSX too, starting with the latest binary, i.e. this one : https://download.tsi.telecom-paristech.fr/gpac/latest_builds/macosx_64/gpac-0.7.2-DEV-rev182-g48308a5-master.dmg

Let me know.

**Comment by bayantaani** [comment]
It still doesn't work, when I play the video from gui (right click -> open with -> osmo4) the app crashes "Osmo4 quit unexpectedly)

**Comment by aureliendavid** [comment]
damn you're right

I tried something else and rebuilt the same package, can you re-download and re-test?

**Comment by bayantaani** [comment]
It's working now. Many thanks! :)

**Comment by bayantaani** [comment]
One more request, can you please update the source code so I can build it from scratch? I want to be able to use the command line options along with the player.

**Comment by aureliendavid** [comment]
There is no code change per say on gpac, what I changed was on the packaging and linking of external libraries. 

If you want to use the command line on mac you can just copy the contents of the dmg and use the binaries. Something like

`cp -rv /Volumes/GPAC\ for\ OSX/Osmo4.app/Contents/MacOS ~/gpac`

Osmo4 is just an old name for MP4Client but the binaries are the same so if you want to use MP4Client as it is called in some documentation you can just copy the binary

`cp ~/gpac/Osmo4 ~/MP4Client`

and then use can use Osmo4/MP4Client as a normal command line program:

```
cd ~/gpac/ 
./MP4Client -guid vid.mp4 -size 600x600 
```


--

If you want to build it from scratch with shvc support you'll have to first build this branch of openhevc : https://github.com/OpenHEVC/openHEVC/tree/HEVC_extensions with something like 

```
cmake -DENABLE_STATIC=ON -DENABLE_EXECUTABLE=OFF -DCMAKE_BUILD_TYPE=RELEASE .
make
```

then copy the built `libLibOpenHevcWrapper.a` library to the `extra_lib/lib/gcc` folder of the gpac source. 

Now when you run `./configure` in gpac you should see 

`OpenHEVCDecoder: yes`

for there you can build gpac (make) with openhevc support.

**Comment by bayantaani** [comment]
I compiled openHEVC, copied the library and ran `./configure` but the support for the decoder is still disabled:
`OpenHEVCDecoder: no`

The directory `extra_lib/lib/gcc` only contains `libLibOpenHevcWrapper.a`
Is there anything else I am missing?

**Comment by aureliendavid** [comment]
you can try to do what configure does, i.e.

create a file ohevc.c with contents

```
#include <stdio.h>
#include <openHevcWrapper.h>
int main( void ) { libOpenHevcInit(1, 1); return 0; }
```

try to compile it with 

`gcc -O3 -Wall -fno-strict-aliasing -Wno-pointer-sign -fPIC -DPIC ohevc.c -Iextra_lib/include -lLibOpenHevcWrapper -lm -lpthread -Wl,--warn-common -Wl,-z,defs -Lextra_lib/lib/gcc`

if there are error messages it should give you an idea of what's missing.

**Comment by bayantaani** [comment]
Thanks! I appreciate the help :)

**Comment by aureliendavid** [comment]
closing for now - reopen if needed

**Comment by sadatUC** [comment]
@bayantaani , Could you please share your configuration file/parameters for encoding (SHVC) a file with 2 layers? I am new to all this and I couldn't find a proper manual for SHVC. I have the HM reference software installed. BUt don't know how to encode a SHVC file.

**Comment by bayantaani** [comment]
Hi @tonmoy42 
I feel you, I also spend a huge amount of time trying out with SHVC
I used SHM12.3, but I believe this applies to most of the recent versions as well (12.x at least). The video I used was originally 4K, and I encoded it into a 1080p base layer and 4K enhancement layer.

I summarized the process in this Github gist:
https://gist.github.com/bayantaani/6976bd9a0547758e40e59c41e7043bd8

All the best!

**Comment by tonmoy42** [comment]
@bayantaani Thank you so much. Can't tell you how grateful I am. I will go through your post. 
Just one more question. Can I have more than two layers? Four would be enough for my purpose. And I just want quality scalability. 
If it's possible to have more than two layers, how would I do that? Just four layers, different bitrate/quality. 

Thanks again!

**Comment by bayantaani** [comment]
I'm glad it helped! :)

I haven't tried more than 2 layers, nor with quality scalability, but I suppose you can.

In your video config file, notice that lines 5-6 describe some parameter called "Level" for each layer. To be honest I don't know what that means, but I think you should add two more lines for layers 3 and 4. Play around with the values and see if anything works.

Moreover, try adding 2 more sets of parameters inside the File I/O section inside the same config file, so that it becomes like this, where (X>Y>Z>W):

#======== File I/O ===============
InputFile0                    : base_layer.yuv
...
QP0                           : X

InputFile1                    : enhancement_layer1.yuv
...
QP1                           : Y   

InputFile2                    : enhancement_layer2.yuv
...
QP2                           : Z

InputFile3                    : enhancement_layer3.yuv
...
QP3                           : W

I assume that you can input the same YUV file to all four layers, since they all have the same resolution.


Also, check the file layers.cfg carefully, the very first line is the number of layers. Set that to 4 and in the same file you will see that each layer should have a set of parameters, try to add the same info for every new layer you want in the output (copy&paste then change some values if necessary, like QP)

Let me know if you find anything interesting!

**Comment by tonmoy42** [comment]
Thanks! I just started. I'll post my updates.

**Comment by tonmoy42** [comment]
@bayantaani  One more question about the configuration. What parameter decides when the quality can be switched? I mean for a 20 sec video, when I play it, can I switch to a higher or lower quality (Ctrl+h, Ctrl+l) at any moment? Or I can only switch after every 40/50/60 frames? How do I set it (quality switch points) up?

**Comment by tonmoy42** [comment]
@bayantaani Thanks to you, I was able to successfully encode two layers with spatial and also two layers with quality (changing QP) scalability. 

Then I tried with four layers as per your instruction, but I see an error like this: https://gist.github.com/tonmoy42/d3ab8d8a43d22fc4a8768c6f383df546

I used the following configuration files along with "encoder_lowdelay_scalable.cfg": 
layers2.cfg: https://gist.github.com/tonmoy42/5733b38aacf66a75e3c4976982eadae8
test2.cfg: https://gist.github.com/tonmoy42/44e8e65d150c36f1a0823614b40f15c0

My .yuv file (352x288): 
https://drive.google.com/file/d/1uy4Q4rJ-dmWR5yprbFuo2377ASIhmjr6/view?usp=sharing

I changed the "Level" for each layer but that didn't remove the error. I saw the description and possible values for the "Level" parameter in the included software-manual.pdf (pg. 9, Table. 4). This is what's written: "Specifies the level to which the encoded bitstream complies. Valid values
are: none, 1, 2, 2.1, 3, 3.1, 4, 4.1, 5, 5.1, 5.2, 6, 6.1, 6.2, 8.5.
NB: There is currently only limited validation that the encoder configuration
complies with the profile, level and tier constraints."

Could you please take a look at it? I couldn't find help anywhere else. It would mean a lot.

**Comment by bayantaani** [comment]
Hi @tonmoy42 I'm sorry for the delayed reply.
I looked at your files and tried them, I got the same errors.
Unfortunately my experience and level of understanding in SHVC is limited to spatial scalability. Try playing around with the config files and see if the error changes.
